### PR TITLE
Added support for requestMtu

### PR DESCRIPTION
**Pull request by pregno**
Since API 21, is possible to set the MTU connection parameter and send packet with length up to 512 bytes.

This implementation bring the support to the RxAndroidBle library.

I've also added the related tests method and updated the sample.
If you prefer I can set up a different activity in the sample for test this functionality.

I've also added the request for the location permission in the sample project.

Tested on a Nexus 5X with Nougat

I know that there is already another PR related to the MTU, but I've implemented as a separated function.

### CODE REVIEW

File: rxandroidble/src/main/java/com/polidea/rxandroidble/RxBleConnection.java [hunk 1]
```diff
@@ -294,4 +294,32 @@ public String toString() {
...
+     * Timeouts after specified amount of time.
+     *
+     * @param timeout multiplier of TimeUnit after which the discovery will timeout in case of no return values
+     * @param timeUnit TimeUnit for the timeout
+     * @return Observable emitting result the MTU requested.
+     * @throws BleGattCannotStartException with {@link BleGattOperationType#ON_MTU_CHANGED} type, when it wasn't possible to set
+     *                                     the MTU for internal reasons.
+     * @throws BleGattException            in case of GATT operation error with {@link BleGattOperationType#ON_MTU_CHANGED} type.
+     */
+    Observable<Integer> requestMtu(int mtu, long timeout, TimeUnit timeUnit);
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/RxBleConnection.java]: I am not really convinced if this functionality is a subject to a timeout - I have not seen any reports of a stalled negotiation of the MTU. It would be best not to have an overcomplicated interface if it is not needed.

**Comment by pregno** [comment]
I've experienced some stalled negotiation when the device is not paired with the phone (or is not paired correctly). 
In that case, sometimes, you will never get a response.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 2]
```diff
@@ -57,6 +58,36 @@ public RxBleConnectionImpl(RxBleRadio rxBleRadio, RxBleGattCallback gattCallback
...
+    public Observable<Integer> requestMtu(int mtu, long timeout, TimeUnit timeUnit) {
+        return privateRequestMtu(mtu, timeout, timeUnit);
+    }
+
+    private Observable<Integer> privateRequestMtu(int mtu, long timeout, TimeUnit timeUnit) {
+        synchronized (discoveredServicesCache) {
+            final Observable<Integer> newObservable;
+            // performing actual discovery
+                newObservable = rxBleRadio
+                        .queue(new RxBleRadioOperationMtuRequest(
```
Comment (uKL) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: Is it possible that the MTU will be changed by the device?

File: sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java [hunk 3]
```diff
@@ -38,6 +45,29 @@ public void onCreate(Bundle savedInstanceState) {
...
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode,
+                                           @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        switch (requestCode) {
+            case PERMISSIONS_REQUEST_LOCATION:
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    // permission was granted, yay! Do the
```
Comment (uKL) [inline sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java]: Our checkstyle cries because of empty statement. In this case it's not relevant but please satisfy it either by adding some action, removing some code or adding checkstyle ignore as this is an example only.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/RxBleConnection.java [hunk 4]
```diff
@@ -294,4 +294,32 @@ public String toString() {
...
+     *
+     * @return Observable emitting result the MTU requested.
+     * @throws BleGattCannotStartException with {@link BleGattOperationType#ON_MTU_CHANGED} type, when it wasn't possible to set
+     *                                     the MTU for internal reasons.
+     * @throws BleGattException            in case of GATT operation error with {@link BleGattOperationType#ON_MTU_CHANGED} type.
+     */
+    Observable<Integer> requestMtu(int mtu);
+
+    /**
+     * Performs GATT request MTU
```
Comment (uKL) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/RxBleConnection.java]: As far as I understand it requires API Level >=21 to use it. I think it would be convenient if the documentation would state that plus there may be an exception like `ApiNotSupportedException(int apiLevel)`. @dariuszseweryn what do you think?

**Comment by dariuszseweryn** [comment]
@pregno could you find some information about the stalled negotiation somewhere on the stackoverflow or other place? I would like to read more about it.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java (hunk 2 above)
Comment (pregno) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: As far as I know, it's not possible.
The device I'm using automatically negotiate the MTU with iOS devices, but on Android this didn't happen

File: sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java (hunk 3 above)
Comment (pregno) [inline sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java]: I don't usually write empty statements :)
I left it empty because I thought was not relevant for the sample.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/RxBleConnection.java (hunk 4 above)
Comment (pregno) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/RxBleConnection.java]: Yes, probably I've put `@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)` in the wrong place.
Right now is in the `protectedRun` method of the `RxBleRadioOperationMtuRequest` class

**Comment by pregno** [comment]
Unfortunately not, but during my tests I've experienced when:
- device not paired
- device appears in the bonded list, but actually is not paired
- device goes out range when connecting

File: sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java (hunk 3 above)
Comment (uKL) [inline sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java]: That's why putting checkstyle ignore would fit in this case :)

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java (hunk 2 above)
Comment (uKL) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: I'm asking because if that's possible then with cache the method may return old values, which is no longer valid.

Comment (pregno) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: In that case we should put a listener (like the connection listener) on the callback `onMtuChanged`.
BTW I'll ask to my firmware developer to test that case

**Comment by dariuszseweryn** [comment]
In this situation how did you tested it? Using what code?

**Comment by pregno** [comment]
The tests has been done on my application, without using your library, so plain (and pain) Android BLE stack.The application is a companion app for our product, so we tested it intensely in every condition.

I know it is a borderline situation, maybe it depends on the firmware implementations.
If you are not convinced about that we could remove the timeout from the request and add it lately if  happens again

**Comment by dariuszseweryn** [comment]
The situation with device going out of range will be automatically handled by the library as `RxBleGattCallback` is reporting an error if the device is disconnected.

About the case when the device is not bonded - I see no hint anywhere that negotiating MTU requires bonding. There are some issues regarding bonding in the Android OS but it's not directly connected with the MTU. If you know more about bonding / mtu - please say or direct me to a place where I can read more.

I think that having a programmable timeout is not the best idea (differently to discovering services as it will be longer if a peripheral has a lot of services / characteristics) because MTU is a two-packet procedure and shouldn't take long. The simpler the API is the better in my opinion.

**Comment by pregno** [comment]
ok

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 5]
```diff
@@ -96,7 +124,8 @@ public RxBleConnectionImpl(RxBleRadio rxBleRadio, RxBleGattCallback gattCallback
     }
 
     @Override
-    public Observable<BluetoothGattCharacteristic> getCharacteristic(@NonNull UUID characteristicUuid) {
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: These formatting changes are not relevant to this commit.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        this.mtu = mtu;
+        this.rxBleGattCallback = rxBleGattCallback;
+        this.bluetoothGatt = bluetoothGatt;
+    }
+
+    @Override
+    protected void protectedRun() {
+        final Subscription subscription = rxBleGattCallback
+                .getOnMtuChanged()
+                .first()
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: Having a `.timeout()` here is not a bad idea - it just doesn't have to be exposed at the interface. Plus a comment about why the timeout is left here could be useful to future maintainers.

File: sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java (hunk 3 above)
Comment (dariuszseweryn) [inline sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java]: This part is also not directly bound to the MTU functionality and could be extracted to a different PR as it will be a benefit even without the MTU.

**Comment by lowwor** [comment]
Can we have a snapshot version of this feature, since it's really useful and I really need this feature.
Or any other way I can achieve this function without compile the whole project as a dependency (It seems BluetoothGatt is not exposed).
Thanks in advance!

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java (hunk 6 above)
Comment (pregno) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: What timeout do you suggest? 20 seconds like for the discovery?

Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: I don't have any BLE device that supports non-standard MTU to test. 20 seconds looks like too much as service discovery is a long task where negotiation should be very quick. If you have a slow device with a less than optimal BLE stack (preferably some cheapish emerging market phone like Micromax Canvas) you could perform some tests how long does it usually take. Could you do the tests?

Comment (pregno) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: Unfortunately I can't test with slow devices. Our device is answering fast, even with the service discovery.

Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: Let's make it 10 seconds then. I think it could be even 5 but can't prove it without tests.

Comment (pregno) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: done

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 7]
```diff
@@ -57,6 +61,33 @@ public RxBleConnectionImpl(RxBleRadio rxBleRadio, RxBleGattCallback gattCallback
     }
 
     @Override
+    public Observable<Integer> requestMtu(int mtu) {
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: Would be good to have this one also annotated with `@RequiresApi`

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 8]
```diff
@@ -57,6 +61,33 @@ public RxBleConnectionImpl(RxBleRadio rxBleRadio, RxBleGattCallback gattCallback
...
+        newObservable = rxBleRadio
+                .queue(new RxBleRadioOperationMtuRequest(
+                        mtu,
+                        gattCallback,
+                        bluetoothGatt,
+                        timeout,
+                        timeUnit,
+                        Schedulers.computation()
+                ))
+                .cacheWithInitialCapacity(1);
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: .cache is not used here basically

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        this.timeoutTimeUnit = timeoutTimeUnit;
+        this.timeoutScheduler = timeoutScheduler;
+    }
+
+    @Override
+    protected void protectedRun() {
+        final Subscription subscription = rxBleGattCallback
+                .getOnMtuChanged()
+                .first()
+                .timeout(timeout, timeoutTimeUnit, timeoutFallbackProcedure(), timeoutScheduler)
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: timeoutFallbackProcedure is not needed here

File: rxandroidble/src/test/groovy/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequestTest.groovy [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        when:
+        changedMtuPublishSubject.onError(testException)
+
+        then:
+        testSubscriber.assertError(testException)
+
+        and:
+        (1.._) * mockSemaphore.release() // technically it's not an error to call it more than once
+    }
+
```
Comment (dariuszseweryn) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequestTest.groovy]: Please add a test for timeout.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/RxBleConnection.java [hunk 11]
```diff
@@ -294,4 +296,17 @@ public String toString() {
      * @return Observable emitting the read RSSI value
      */
     Observable<Integer> readRssi();
+
+    /**
+     * Performs GATT request MTU
+     *
+     * Timeouts after specified amount of time.
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/RxBleConnection.java]: Timeouts after 10 seconds.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleGattCallback.java [hunk 12]
```diff
@@ -12,8 +12,8 @@
 import com.polidea.rxandroidble.exceptions.BleGattException;
 import com.polidea.rxandroidble.exceptions.BleGattOperationType;
 import com.polidea.rxandroidble.internal.RxBleLog;
-
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleGattCallback.java]: not relevant to the functionality

**Comment by dariuszseweryn** [comment]
Thanks for re-adding the timeout as it is useful till we will deal with bonding related issues. I have added comments to the code.
Things to do before I can merge the changes:

1. Fix checkstyle issues.
2. Add a test for timeouting MTU negotiation operation.
3. Add a `RequiresApi` on `RxBleConnectionImpl.requestMtu(int mtu)` call.
4. Remove all changes that do not touch functionality of the mtu. (new lines in import).
5. Extract changes regarding getting the permissions to a different pull request.

Once the above things will be addressed we should be there. 👍

File: rxandroidble/src/test/groovy/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequestTest.groovy (hunk 10 above)
Comment (pregno) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequestTest.groovy]: done

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleGattCallback.java (hunk 12 above)
Comment (pregno) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleGattCallback.java]: what is not relevant?

Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleGattCallback.java]: removed line :) And added one line below

Comment (pregno) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleGattCallback.java]: oh c'mon man, Android studio added

Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleGattCallback.java]: From my experience any added / deleted line that is not directly connected to the change only obfuscates the code. Minimal changes are the best changes as it is easier to spot mistakes.

**Comment by dariuszseweryn** [comment]
Three things left:

1. There are some obsolete changes with line breaks in `RxBleConnectionImpl`
2. Checkstyle is whining about unused imports in `RxBleRadioOperationMtuRequest` you can check it here: https://travis-ci.org/Polidea/RxAndroidBle/builds/181702296
3. There is now unused line in `ScanActivity` -> `public final static int PERMISSIONS_REQUEST_LOCATION = 16;`

I will mark them in the code to make it easier.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 13]
```diff
@@ -107,7 +135,8 @@ public RxBleConnectionImpl(RxBleRadio rxBleRadio, RxBleGattCallback gattCallback
     }
 
     @Override
-    public Observable<Observable<byte[]>> setupNotification(@NonNull BluetoothGattCharacteristic characteristic) {
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: this change with line break is obsolete

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 14]
```diff
@@ -117,7 +146,8 @@ public RxBleConnectionImpl(RxBleRadio rxBleRadio, RxBleGattCallback gattCallback
     }
 
     @Override
-    public Observable<Observable<byte[]>> setupIndication(@NonNull BluetoothGattCharacteristic characteristic) {
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: and this change with line break is obsolete

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 15]
```diff
@@ -225,7 +255,8 @@ private void dismissTriggeredRead(
     }
 
     @Override
-    public Observable<byte[]> readCharacteristic(@NonNull BluetoothGattCharacteristic characteristic) {
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: and this change with line break is obsolete

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 16]
```diff
@@ -234,20 +265,23 @@ private void dismissTriggeredRead(
     }
 
     @Override
-    public Observable<byte[]> writeCharacteristic(@NonNull UUID characteristicUuid, @NonNull byte[] data) {
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: and this change with line break is obsolete

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 17]
```diff
@@ -234,20 +265,23 @@ private void dismissTriggeredRead(
...
+    public Observable<byte[]> writeCharacteristic(
+            @NonNull UUID characteristicUuid, @NonNull byte[] data) {
         return getCharacteristic(characteristicUuid)
                 .flatMap(characteristic -> writeCharacteristic(characteristic, data));
     }
 
     @Deprecated
     @Override
-    public Observable<BluetoothGattCharacteristic> writeCharacteristic(@NonNull BluetoothGattCharacteristic bluetoothGattCharacteristic) {
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: and this change with line break is obsolete

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 18]
```diff
@@ -234,20 +265,23 @@ private void dismissTriggeredRead(
...
+    public Observable<BluetoothGattCharacteristic> writeCharacteristic(
+            @NonNull BluetoothGattCharacteristic bluetoothGattCharacteristic) {
         return writeCharacteristic(bluetoothGattCharacteristic, bluetoothGattCharacteristic.getValue())
                 .map(bytes -> bluetoothGattCharacteristic);
     }
 
     @Override
-    public Observable<byte[]> writeCharacteristic(@NonNull BluetoothGattCharacteristic characteristic, @NonNull byte[] data) {
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: and this change with line break is obsolete

File: sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java [hunk 19]
```diff
@@ -23,6 +28,8 @@
 
 public class ScanActivity extends AppCompatActivity {
 
+    public final static int PERMISSIONS_REQUEST_LOCATION = 16;
```
Comment (dariuszseweryn) [inline sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java]: line 31 and 32 are not needed as the permission request was extracted to a different PR

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java [hunk 20]
```diff
@@ -0,0 +1,65 @@
+package com.polidea.rxandroidble.internal.operations;
+
+import android.bluetooth.BluetoothGatt;
+import android.os.Build;
+import android.support.annotation.NonNull;
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: This import is unused according to https://travis-ci.org/Polidea/RxAndroidBle/builds/181702296 checkstyle.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.polidea.rxandroidble.exceptions.BleGattCannotStartException;
+import com.polidea.rxandroidble.exceptions.BleGattOperationType;
+import com.polidea.rxandroidble.internal.RxBleRadioOperation;
+import com.polidea.rxandroidble.internal.connection.RxBleGattCallback;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import rx.Observable;
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: This import is unused according to https://travis-ci.org/Polidea/RxAndroidBle/builds/181702296 checkstyle.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+import android.support.annotation.NonNull;
+import android.support.annotation.RequiresApi;
+
+import com.polidea.rxandroidble.exceptions.BleGattCannotStartException;
+import com.polidea.rxandroidble.exceptions.BleGattOperationType;
+import com.polidea.rxandroidble.internal.RxBleRadioOperation;
+import com.polidea.rxandroidble.internal.connection.RxBleGattCallback;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/RxBleRadioOperationMtuRequest.java]: This import is unused according to https://travis-ci.org/Polidea/RxAndroidBle/builds/181702296 checkstyle.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 23]
```diff
@@ -4,17 +4,21 @@
...
 import android.bluetooth.BluetoothGattService;
+import android.os.Build;
 import android.support.annotation.NonNull;
+import android.support.annotation.RequiresApi;
 
 import com.polidea.rxandroidble.RxBleConnection;
 import com.polidea.rxandroidble.RxBleDeviceServices;
 import com.polidea.rxandroidble.exceptions.BleCannotSetCharacteristicNotificationException;
 import com.polidea.rxandroidble.exceptions.BleConflictingNotificationAlreadySetException;
+import com.polidea.rxandroidble.exceptions.BleException;
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: This import is unused according to https://travis-ci.org/Polidea/RxAndroidBle/builds/181702296 checkstyle.

File: rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java [hunk 24]
```diff
@@ -57,6 +61,30 @@ public RxBleConnectionImpl(RxBleRadio rxBleRadio, RxBleGattCallback gattCallback
...
     @Override
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    public Observable<Integer> requestMtu(int mtu) {
+        return privateRequestMtu(mtu, 10, TimeUnit.SECONDS);
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private Observable<Integer> privateRequestMtu(int mtu, long timeout, TimeUnit timeUnit) {
+        final Observable<Integer> newObservable;
+        // performing actual discovery
```
Comment (dariuszseweryn) [inline rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleConnectionImpl.java]: this comment is not needed (probably a copy-paste error)

**Comment by dariuszseweryn** [comment]
@pregno I hope that you are not taking my requests personally. Our current setup is currently not building artifacts if the `checkstyle` is complaining - thus the need of removing unused imports. And it doesn't look good if there are changes that are not relevant to the topic of a pull request - so removing unnecessary formatting and leftovers from parts that were already removed will boil down the code to easily understandable bare minimum. 
I am really thankful to you for committing your work towards what I hope will be the best android ble library. :+1:

**Comment by pregno** [comment]
Don't worry :)
I will close the various fixes soon

File: sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java [hunk 25]
```diff
@@ -23,6 +23,7 @@
 
 public class ScanActivity extends AppCompatActivity {
 
+
```
Comment (dariuszseweryn) [inline sample/src/main/java/com/polidea/rxandroidble/sample/example1_scanning/ScanActivity.java]: this is not needed

**Comment by pregno** [comment]
ok, done

**Comment by dariuszseweryn** [comment]
Thank you for your effort @pregno - your changes are merged and will be available in the 1.2.0-SNAPSHOT release shortly. :) 💯 
Best Regards

**Comment by pregno** [comment]
thank you too!

best

**Comment by lowwor** [comment]
Great job! Thank you all for this nice feature.
